src/App.tsx

```diff
@@ -11,18 +11,32 @@ import { OnboardContext } from "common/context/OnboardContext";
 import { recoverPublicKey } from "./features/vote/helpers/recoverPublicKey";
 import { derivePrivateKey } from "./features/vote/helpers/derivePrivateKey";
 import { useCurrentRoundId } from "hooks";
+import web3 from "web3";
+import has from "lodash.has";
+import setWith from "lodash.setwith";
+
 interface Props {
   queryClient: QueryClient;
 }
 
+export interface SigningKey {
+  publicKey: string;
+  privateKey: string;
+  roundMessage: string;
+  roundId: string;
```

Comment:
Two questions:
1. Why do we need these variables (roundId and roundMessage) here if they're effectively encoded into the mapping, itself?
2. Why do we need a concept of round hardcoded into the data structure? It seems like we could be agnostic to how the message is generated or on what frequency it changes and leave that up to the message generator, no?

Reply:
It totally makes sense to store them, but it's unclear why they need to be associated with the signing key. Instead of doing checks on this data, isn't the more standard pattern to have a useMemo or useEffect that sets the computed message (or roundId) as a dependency?

Something like (oversimplified):
```ts
export interface SigningKey {
  publicKey: string;
  privateKey: string;
};

// Note: connectionStatus is some variable that changes when we the user changes something about their wallet connection.

useEffect(() => {
  // Compute message
  const message = getMessage(roundId);
  updateSigningKey(message);
}, [connectionStatus, roundId]);

```

The reason I bring this up is that I think the generation of the message and how we store it should ideally be completely abstracted away from one another. The storage should only care about mapping from message to keys, with no regard to how that message is generated. We only need to re-evaluate our signing keys when some input (either the signing address, connection status, or the message) changes.

Does that make sense? Is my understanding of the fundamental logic wrong?

Reply:
It makes sense, but I am trying to think of how we'd organize this. Because it makes sense to keep both of these pieces of data close and in memory, as it's essentially part of the user's logged in state.

It helps because if the user is on the app when the round crosses over, we need to make sure we are doing some direct comparisons.